pkg/reconciler/apis/apiexportendpointsliceurls/apiexportendpointsliceurls_controller.go

```diff
@@ -0,0 +1,481 @@
+
+	apiExportEndpointSliceClusterInformer       apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+	globalApiExportEndpointSliceClusterInformer apisv1alpha1informers.APIExportEndpointSliceClusterInformer
+}
+
+func (c *controller) enqueueAPIExportEndpointSliceByAPIBinding(obj interface{}) {
+	if d, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = d.Obj
+	}
+	binding, ok := obj.(*apisv1alpha1.APIBinding)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("obj is supposed to be a APIBinding, but is %T", obj))
+		return
+	}
+
+	{ // local to shard
+		keys := sets.New[string]()
+		if path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path); !path.Empty() {
+			pathKeys, err := c.apiExportEndpointSliceClusterInformer.Informer().GetIndexer().IndexKeys(indexers.APIExportEndpointSliceByAPIExport, path.Join(binding.Spec.Reference.Export.Name).String())
+			if err != nil {
+				utilruntime.HandleError(err)
+				return
+			}
+			keys.Insert(pathKeys...)
+		}
+
+		for _, key := range sets.List[string](keys) {
+			slice, exists, err := c.apiExportEndpointSliceClusterInformer.Informer().GetIndexer().GetByKey(key)
+			if err != nil {
+				utilruntime.HandleError(err)
+				continue
+			} else if !exists {
+				continue
+			}
+			logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*apisv1alpha1.APIBinding))
+			logging.WithQueueKey(logger, key).V(4).Info("queuing APIExportEndpointSlices because of consumer APIBinding")
+			c.enqueueAPIExportEndpointSlice(slice)
+		}
+	}
+	{
+		keys := sets.New[string]()
+		if path := logicalcluster.NewPath(binding.Spec.Reference.Export.Path); !path.Empty() {
+			pathKeys, err := c.globalApiExportEndpointSliceClusterInformer.Informer().GetIndexer().IndexKeys(indexers.APIExportEndpointSliceByAPIExport, path.Join(binding.Spec.Reference.Export.Name).String())
+			if err != nil {
+				utilruntime.HandleError(err)
+				return
+			}
+			keys.Insert(pathKeys...)
+		}
+
+		for _, key := range sets.List[string](keys) {
+			slice, exists, err := c.globalApiExportEndpointSliceClusterInformer.Informer().GetIndexer().GetByKey(key)
+			if err != nil {
+				utilruntime.HandleError(err)
+				continue
+			} else if !exists {
+				continue
+			}
+			logger := logging.WithObject(logging.WithReconciler(klog.Background(), ControllerName), obj.(*apisv1alpha1.APIBinding))
+			logging.WithQueueKey(logger, key).V(4).Info("queuing APIExportEndpointSlices because of consumer APIBinding from cache")
+			c.enqueueAPIExportEndpointSlice(slice)
+		}
+	}
+}
+
+// enqueueAPIExportEndpointSlice enqueues an APIExportEndpointSlice.
+func (c *controller) enqueueAPIExportEndpointSlice(obj interface{}) {
+	key, err := kcpcache.DeletionHandlingMetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(4).Info("queueing APIExportEndpointSlice")
```

Comment:
can we add a suffix argument to the method and then do `logger.V(4).Info("queueing APIExportEndpointSlice"+suffix)` to get `queueing APIExportEndpointSlice because APIExport X` (we have some examples already), or use contextual logging with `reason` keys.